brainframe_qt/api_utils/streaming/stream_manager.py

```diff
@@ -1,139 +1,175 @@
-            otherwise
+        self._running_streams: List[int] = []
+        """Currently running streams. Does not include stay-alive streams.
+        
+        Max length should be _NUM_ACTIVE_STREAMS
         """
-        return stream_id in self._stream_readers
+        self._paused_streams: List[int] = []
+        """Currently paused streams"""
 
-    def close_stream(self, stream_id: int) -> None:
-        """Close a specific stream and remove the reference.
+        self.stream_readers: Dict[int, SyncedStreamReader] = {}
+        """All StreamReaders currently instantiated, paused or unpaused"""
 
-        :param stream_id: The ID of the stream to delete
-        """
-        stream = self.close_stream_async(stream_id)
-        stream.wait_until_closed()
+        self._init_signals()
+
+    def _init_signals(self) -> None:
+        self.destroyed.connect(self.close)
 
     def close(self) -> None:
-        """Close all streams and remove references"""
-        for stream_id in self._stream_readers.copy().keys():
-            self.close_stream_async(stream_id)
-        self._stream_readers = {}
-
-        for stream in self._async_closing_streams:
-            stream.wait_until_closed()
-            self._async_closing_streams.remove(stream)
-
-    def close_stream_async(self, stream_id: int) -> SyncedStreamReader:
-        stream = self._stream_readers.pop(stream_id)
-        self._async_closing_streams.append(stream)
-        stream.close()
-        return stream
-
-
-class StreamManagerAPI(BrainFrameAPI):
-    """Augments the API class to manage and provide a StreamManager."""
-
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-        self._stream_manager = None
-
-    def delete_stream_configuration(self, stream_id,
-                                    timeout=120):
-        super().delete_stream_configuration(stream_id, timeout)
-        if self._stream_manager is not None \
-                and self._stream_manager.is_streaming(stream_id):
-            self._stream_manager.close_stream_async(stream_id)
-
-    def get_stream_manager(self):
-        """Returns a singleton StreamManager object"""
-        # Lazily import streaming code to avoid OpenCV dependencies unless
-        # necessary
-        from brainframe_qt.api_utils.streaming import StreamManager
-
-        if self._stream_manager is None:
-            self._stream_manager = StreamManager(self.get_status_receiver())
-        return self._stream_manager
-
-    def get_stream_reader(self, stream_config: StreamConfiguration):
-        """Get the SyncedStreamReader for the given stream_configuration.
-
-        :param stream_config: The stream configuration to open.
-        :return: A SyncedStreamReader object
+        """Request and wait for all streams to close"""
+        logging.info("Initiating StreamManager close")
+        self._close()
+        logging.info("StreamManager close finished")
```

Comment:
Could `close()` and `_close()` be combined?

Reply:
All the other functions have internal/external functions, so I figured it was good to keep the design consistent